### PR TITLE
Lt 1320 view ecju responses

### PR DESCRIPTION
**Pull request by BenAndrew-Kainos**

### CODE REVIEW

File: ui_automation_tests/features/ecju_query.feature [hunk 1]
```diff
@@ -30,3 +30,21 @@ Feature: I want to create ECJU queries
     Then the new ECJU Query is visible in the list
     When I click back
     Then the ECJU Query creation is visible in the case timeline
+
```
Comment (phlptt) [inline ui_automation_tests/features/ecju_query.feature]: two lines here

File: ui_automation_tests/features/ecju_query.feature [hunk 2]
```diff
@@ -30,3 +30,21 @@ Feature: I want to create ECJU queries
...
+  @LT_1320_view_response_to_ecju
+  Scenario: Add an ECJU Query and respond and view response
+    Given I create application or application has been previously created
+    And I sign in to SSO or am signed into SSO
+    And I create an ecju query picklist
+    When I go to application previously created
```
Comment (phlptt) [inline ui_automation_tests/features/ecju_query.feature]: And not when

File: ui_automation_tests/features/ecju_query.feature [hunk 3]
```diff
@@ -30,3 +30,21 @@ Feature: I want to create ECJU queries
...
+    And I click the ECJU Queries button
+    And I click Add an ECJU Query
+    And I Select Write a new question
+    And I click continue
+    And I enter text in the question text area
+    And I click continue
+    And I click Yes
+    And I click continue
```
Comment (phlptt) [inline ui_automation_tests/features/ecju_query.feature]: http://itsadeliverything.com/declarative-vs-imperative-gherkin-scenarios-for-cucumber

We should be using a declarative style not imperative - have a read of the page above . (not for this story for future stories)

File: ui_automation_tests/helpers/helpers.py [hunk 4]
```diff
@@ -153,22 +152,12 @@ def get_element_index_by_partial_text(elements, text: str):
...
 
-def wait_until_menu_is_visible(driver):
-    while True:
-        try:
-            if driver.find_element_by_css_selector('.lite-menu--visible').is_displayed():
-                break
-        except Exception: # noqa
-            continue
+def menu_is_visible(driver):
+    return driver.find_element_by_css_selector('.lite-menu--visible').is_displayed()
```
Comment (phlptt) [inline ui_automation_tests/helpers/helpers.py]: have you tried thuis when it is not displayed ? Im pretty sure itll error out the whole test

Comment (phlptt) [reply ui_automation_tests/helpers/helpers.py]: as is_displayed either returns True or error

File: ui_automation_tests/helpers/seed_data.py [hunk 5]
```diff
@@ -367,6 +368,21 @@ def assign_test_cases_to_bin(self, bin_queue_id, new_cases_queue_id):
...
+    def add_ecju_response(self, question, response):
+        self.log("adding response to ecju: ...")
+        case_id = self.context['case_id']
+        ecju_queries = self.make_request("GET", url='/cases/' + case_id + '/ecju-queries/', headers=self.gov_headers)
+        ecju_query_id = None
+        for ecju_query in ecju_queries.json()['ecju_queries']:
+            if ecju_query['question'] == question:
+                ecju_query_id = ecju_query['id']
+                break
+        assert ecju_query_id
```
Comment (phlptt) [inline ui_automation_tests/helpers/seed_data.py]: as we're doing asserts in seed data wchih i dont feel too comfortable about i think we should at least have a failure message here to say "failed to add ecju response via seed data, test going to finish.."
or something like this

File: cases/views/main.py [hunk 6]
```diff
@@ -10,7 +10,7 @@
 from s3chunkuploader.file_handler import S3FileUploadHandler, s3_client
 
 from cases.forms.attach_documents import attach_documents_form
-from cases.forms.create_ecju_query import create_ecju_query_write_or_edit_form, choose_ecju_query_type_form, \
+from cases.forms.create_ecju_query import choose_ecju_query_type_form, create_ecju_query_write_or_edit_form, \
```
Comment (stephanieboyd) [inline cases/views/main.py]: Need to remove the rest of what's been refactored into `views/ecju.py`

File: ui_automation_tests/pages/case_list_page.py [hunk 7]
```diff
@@ -33,21 +35,24 @@ class CaseListPage(BasePage):
...
+        if number_of_pages != 0:
+            while is_present == 0:
+                url = self.driver.current_url
+                if 'page' not in url:
+                    self.driver.find_element_by_id("page-" + str(number_of_pages)).click()
+                else:
+                    next_page = url + '&page=' + str(page_number)
+                    self.driver.get(next_page)
+                page_number -= 1
+                is_present = len(self.driver.find_elements_by_link_text(id))
```
Comment (phlptt) [inline ui_automation_tests/pages/case_list_page.py]: 👍

File: ui_automation_tests/step_defs/test_queues.py [hunk 8]
```diff
@@ -49,6 +46,7 @@ def see_queue_in_queue_list(driver, context):
 
 @then('I dont see previously created application')
 def dont_see_queue_in_queue_list(driver, context):
+    driver.refresh()
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_queues.py]: dont think this refresh is needed

File: cases/views/ecju.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+from lite_forms.generators import form_page, error_page
+
+from cases.forms.create_ecju_query import choose_ecju_query_type_form, create_ecju_query_write_or_edit_form, \
+    create_ecju_create_confirmation_form
+from cases.services import get_ecju_queries, post_ecju_query
+from picklists.services import get_picklists, get_picklist_item
+
+
+class ViewEcjuQueries(TemplateView):
+    def _get_ecju_queries(self, request, case_id):
```
Comment (lukesa) [inline cases/views/ecju.py]: why you start method name from underscore?

File: cases/views/ecju.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        open_ecju_queries = list()
+        closed_ecju_queries = list()
+        for query in ecju_queries.get('ecju_queries'):
+            if query.get('response'):
+                closed_ecju_queries.append(query)
+            else:
+                open_ecju_queries.append(query)
+        return open_ecju_queries, closed_ecju_queries
+
+    def get(self, request, **kwargs):
```
Comment (lukesa) [inline cases/views/ecju.py]: add doctoring

File: cases/views/ecju.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            'open_ecju_queries': open_ecju_queries,
+            'closed_ecju_queries': closed_ecju_queries
+        }
+        return render(request, 'cases/case/ecju-queries.html', context)
+
+
+class CreateEcjuQuery(TemplateView):
+    NEW_QUESTION_DDL_ID = 'new_question'
+
+    def get(self, request, **kwargs):
```
Comment (lukesa) [inline cases/views/ecju.py]: add docstring

File: cases/views/ecju.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        picklist_choices = [Option(self.NEW_QUESTION_DDL_ID, 'Write a new question')] + \
+                           [Option(picklist.get('id'), picklist.get('name')) for picklist in picklists]
+        form = choose_ecju_query_type_form(
+            reverse('cases:ecju_queries', kwargs={'pk': case_id}),
+            picklist_choices
+        )
+
+        return form_page(request, form, extra_data={'case_id': case_id})
+
+    def post(self, request, **kwargs):
```
Comment (lukesa) [inline cases/views/ecju.py]: add docstring

File: ui_automation_tests/helpers/wait.py [hunk 13]
```diff
@@ -1,6 +1,8 @@
 import time
 
-# How many attempts to wait for the function to return True
+# How long in seconds the function should be attempted until giving up
```
Comment (stephanieboyd) [inline ui_automation_tests/helpers/wait.py]: Move the comment down a line, so it's above what it's referring to

File: ui_automation_tests/helpers/helpers.py (hunk 4 above)
Comment (BenAndrew-Kainos) [reply ui_automation_tests/helpers/helpers.py]: As discussed we'll leave this out for now but could be a consideration if this starts failing

File: ui_automation_tests/step_defs/test_queues.py (hunk 8 above)
Comment (stephanieboyd) [inline ui_automation_tests/step_defs/test_queues.py]: Just a question - why you need this now?

File: ui_automation_tests/features/ecju_query.feature (hunk 3 above)
Comment (BenAndrew-Kainos) [reply ui_automation_tests/features/ecju_query.feature]: Good to know for the future

File: cases/views/ecju.py (hunk 9 above)
Comment (BenAndrew-Kainos) [reply cases/views/ecju.py]: I didn't write this method but I believe it's a naming convention for private methods

File: ui_automation_tests/step_defs/test_queues.py (hunk 8 above)
Comment (stephanieboyd) [reply ui_automation_tests/step_defs/test_queues.py]: I hadn't refreshed - I see Nick's comment :)

File: cases/views/ecju.py (hunk 10 above)
Comment (BenAndrew-Kainos) [reply cases/views/ecju.py]: I didn't write this method but will give it a shot

File: cases/views/ecju.py (hunk 11 above)
Comment (BenAndrew-Kainos) [reply cases/views/ecju.py]: Same as above

File: cases/views/ecju.py (hunk 12 above)
Comment (BenAndrew-Kainos) [reply cases/views/ecju.py]: Same as above

File: cases/services.py [hunk 14]
```diff
@@ -99,60 +105,18 @@ def post_case_advice(request, case_pk, json):
...
+                      'countries': 'country', 'goods': 'good', 'goods_types': 'good_type'}
+
+    for entity_name in single_cases:
+        if json.get(entity_name):
+            new_data.append(build_case_advice(entity_name, json.get(entity_name), base_data))
+
+    for entity_name, entity_name_singular in multiple_cases.items():
+        if json.get(entity_name):
+            for entity in json.get(entity_name, []):
+                new_data.append(build_case_advice(entity_name_singular, entity, base_data))
```
Comment (andytaylorkainos) [inline cases/services.py]: Nice refactor!

File: cases/urls.py [hunk 15]
```diff
@@ -33,9 +33,9 @@
     # ex: /cases/<uuid:pk>/advice-view/give-advice/approve/
     path('<uuid:pk>/advice-view/give-advice/<str:type>/', advice.GiveAdviceDetail.as_view(), name='give_advice_detail'),
     # ex: /cases/<uuid:pk>/ecju-queries/
-    path('<uuid:pk>/ecju-queries/', main.ViewEcjuQueries.as_view(), name='ecju_queries'),
+    path('<uuid:pk>/ecju-queries/', ecju.ViewEcjuQueries.as_view(), name='ecju_queries'),
```
Comment (andytaylorkainos) [inline cases/urls.py]: 👍

File: cases/views/ecju.py (hunk 9 above)
Comment (andytaylorkainos) [reply cases/views/ecju.py]: method names prefixed with `_` are used to note that they are intended to be used as private methods

File: cases/views/ecju.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        case_id = str(kwargs['pk'])
+        form_name = request.POST.get('form_name')
+
+        if form_name == 'ecju_query_type_select':
+            return self._handle_ecju_query_type_select_post(request, case_id)
+
```
Comment (andytaylorkainos) [inline cases/views/ecju.py]: can we remove the empty lines between each `if-else` please

File: cases/views/ecju.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+        elif form_name == 'ecju_query_write_or_edit_question':
+            return self._handle_ecju_query_write_or_edit_post(case_id, request)
+
+        elif form_name == 'ecju_query_create_confirmation':
+            return self._handle_ecju_query_confirmation_post(case_id, request)
+
+        else:
+            # Submitted data does not contain an expected form field - return an error
```
Comment (andytaylorkainos) [inline cases/views/ecju.py]: 👍

File: cases/views/ecju.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+            return self._handle_ecju_query_form_errors(case_id, ecju_query, request)
+        else:
+            form = create_ecju_create_confirmation_form()
+            form.questions.append(HiddenField('question', request.POST.get('question')))
+            return form_page(request, form)
+
+    def _handle_ecju_query_confirmation_post(self, case_id, request):
+        data = {'question': request.POST.get('question')}
+
+        if request.POST.get('ecju_query_confirmation') == 'yes':
```
Comment (andytaylorkainos) [inline cases/views/ecju.py]: should we add `.lower()` to this evaluation? just a thought - not enforcing

File: cases/views/ecju.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+            ecju_query, status_code = post_ecju_query(request, case_id, data)
+
+            if status_code != 201:
+                return self._handle_ecju_query_form_errors(case_id, ecju_query, request)
+            else:
+                return redirect(reverse('cases:ecju_queries', kwargs={'pk': case_id}))
+        elif request.POST.get('ecju_query_confirmation') == 'no':
```
Comment (andytaylorkainos) [inline cases/views/ecju.py]: same as before (only suggestion)

File: templates/cases/case/ecju-queries.html [hunk 20]
```diff
@@ -15,16 +15,41 @@ <h2 class="govuk-heading-l">{% get_string 'cases.ecju_queries.title' %}</h2>
...
-	        <div class="lite-table-row">
-	            <div class="lite-table__cell">
-	                {{ query.question }}
-	            </div>
-	        </div>
-		{% endfor %}
-	    </div>
+	{% if open_ecju_queries or closed_ecju_queries %}
+        {% if open_ecju_queries %}
+            <h3 class="govuk-heading-m">Open queries</h3>
```
Comment (andytaylorkainos) [inline templates/cases/case/ecju-queries.html]: Maybe a lite-content string?

File: ui_automation_tests/pages/case_list_page.py [hunk 21]
```diff
@@ -112,13 +105,9 @@ def click_on_queue_title(self):
...
     def click_on_queue_name(self, queue_name):
         self.click_on_queue_title()
-        elements = self.driver.find_elements_by_css_selector(self.dropdown_item)
-        for idx, element in enumerate(elements):
-            if queue_name in element.text:
-                self.driver.execute_script(
-                    "document.getElementsByClassName('" + self.dropdown_item_class + "')[" + str(idx) + "].scrollIntoView(true);")
-                element.click()
-                break
+        time.sleep(.5)
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply ui_automation_tests/pages/case_list_page.py]: Being pernickety, can we append the `.5` with `0` :)

File: ui_automation_tests/pages/case_list_page.py [hunk 22]
```diff
@@ -112,13 +105,9 @@ def click_on_queue_title(self):
...
+        self.driver.execute_script("document.getElementById('" + queue_name + "').scrollIntoView(true);")
```
Comment (andytaylorkainos) [inline ui_automation_tests/pages/case_list_page.py]: did we not have a helper method for scrolling into view? (if not; hint hint)

File: ui_automation_tests/pages/case_list_page.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply ui_automation_tests/pages/case_list_page.py]: They fail otherwise

File: cases/views/ecju.py (hunk 18 above)
Comment (BenAndrew-Kainos) [reply cases/views/ecju.py]: I don't think so, it's set as a value in the Option lite-forms element

Comment (BenAndrew-Kainos) [reply cases/views/ecju.py]: Done anyway

**Pull request merged**